### PR TITLE
fix SES API limits

### LINKED ISSUE #9644: AWS SES Transport fails on sending a batch immediately
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
The more details the better...

| Q   | A
| --- | ---
| Mautic version | 3.2.5
| PHP version | 7.4
| Browser | 

### Steps to reproduce
1. Make sure you are sending using immediate mode 
2. Create a segment with more than 50 contacts
3. Send an email to that segment
4. The email will be failed to send
5. There is a related error for throttling (logs below) 
 
### Log errors
```
[2021-02-03 08:01:33] mautic.ERROR: [MAIL ERROR] Token amount (51) is larger than the capacity (3). (send); {"exception":"[object] (Swift_TransportException(code: 0): Token amount (51) is larger than the capacity (3). at app/bundles/EmailBundle/Swiftmailer/Transport/AbstractTokenArrayTransport.php:300)"} []

[2021-02-05 08:25:00] mautic.ERROR: [MAIL ERROR] Expected response code 250 but got code "454", with message "454 Throttling failure: Maximum sending rate exceeded. " (send); email@domain.com {"exception":"[object] (Swift_TransportException(code: 454): Expected response code 250 but got code \"454\", with message \"454 Throttling failure: Maximum sending rate exceeded.\r\n\" at /home/smartconnectamba/public_html/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Transport/AbstractSmtpTransport.php:457)"} []
```

### Solution to be implemented must look into the following lines: 
```
        if (!$this->started) {
            $this->amazonClient  = $this->createAmazonClient();

            $account             = $this->amazonClient->getAccount();
            $emailQuotaRemaining = $account->get('SendQuota')['Max24HourSend'] - $account->get('SendQuota')['SentLast24Hours'];

            if (!$account->get('SendingEnabled')) {
                $this->logger->error('Your AWS SES is not enabled for sending');
                throw new \Exception('Your AWS SES is not enabled for sending');
            }

            if (!$account->get('ProductionAccessEnabled')) {
                $this->logger->info('Your AWS SES is in sandbox mode, consider moving it to production state');
            }

            if ($emailQuotaRemaining <= 0) {
                $this->logger->error('Your AWS SES quota is currently exceeded, used '.$account->get('SentLast24Hours').' of '.$account->get('Max24HourSend'));
                throw new \Exception('Your AWS SES quota is currently exceeded');
            }

            $this->concurrency   = floor($account->get('SendQuota')['MaxSendRate']);

            $this->started = true;
        }
```

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by mabumusa1**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" 
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #9644, #9627, #9887

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Make sure your sending in the immediate mode 
3. Try to send a campaign of 500 emails 
4. API limits should be respected and no errors should show in the logs and you should not exceed the limits of SES

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9735"></a>

### CODE REVIEW

**Comment by mabumusa1** [comment]
@RCheesley who would you suggest to review this PR? 

I tested locally and it worked as expected, I had an if statement that prevented the code from working as expected

**Comment by RCheesley** [comment]
@mabumusa1 probably someone who is using it in production would be good, and @mautic/core-team for a code review.

**Comment by RCheesley** [comment]
PS automated tests need some ❤️

**Comment by RCheesley** [comment]
Also have changed the base as this is a bug fix so will likely need to target the 3.3 branch - may need to be rebased.

**Comment by mabumusa1** [comment]
> Also have changed the base as this is a bug fix so will likely need to target the 3.3 branch - may need to be rebased.

Done, and test suite included all the cases

**Comment by RCheesley** [comment]
@Wise-Gorilla are you able to test this one so that we could consider it for the 3.3.2 Release Candidate on Monday?

File: app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php [hunk 1]
```diff
@@ -265,6 +263,7 @@ public function send(\Swift_Mime_SimpleMessage $toSendMessage, &$failedRecipient
                     }
                 },
                 'rejected' => function (AwsException $reason, $iteratorId) use ($evt) {
+                    dd($reason);
```
Comment (kuzmany) [inline app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php]: Looks like test code, isn't it?

**Review by kuzmany — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php (hunk 1 above)
Comment (mabumusa1) [reply app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php]: That is correct, I think I missed it. I just removed this line

**Comment by mabumusa1** [comment]
@dennisameling I am facing a problem in the tests, I think I've seen this issue before, did you merge the PR to fix this issue or not yet?

**Comment by dennisameling** [comment]
@mabumusa1 I ran into this as well but don't have a clue what the problem or the solution is here. Hope to have a look in the coming weekend

**Comment by do-web** [comment]
@RCheesley  Any updates here when this will be released?

**Comment by dennisameling** [comment]
@do-web Could you help us by testing if this PR fixes the issue for you? CI is green now, so we only need testers to confirm it's indeed been fixed, and a code review

**Comment by do-web** [comment]
> @do-web Could you help us by testing if this PR fixes the issue for you? CI is green now, so we only need testers to confirm it's indeed been fixed, and a code review

@dennisameling How can i test?

**Comment by dennisameling** [comment]
@do-web I guess the easiest is as follows. Can you access a command line for your Mautic installation?

```
(First go to the folder where you have installed Mautic!)
cd app/bundles/EmailBundle/Swiftmailer/Transport
mv AmazonApiTransport.php AmazonApiTransport.bak.php
wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php

(First go to the folder where you have installed Mautic!)
cd app/bundles/EmailBundle/Translations/en_US
mv messages.ini messages.bak.ini
wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Translations/en_US/messages.ini
```

Then test if it fixes the problem.
If you want to revert back to the old code for whatever reason, simply run:

```
(First go to the folder where you have installed Mautic!)
cd app/bundles/EmailBundle/Swiftmailer/Transport
rm AmazonApiTransport.php
mv AmazonApiTransport.bak.php AmazonApiTransport.php

(First go to the folder where you have installed Mautic!)
cd app/bundles/EmailBundle/Translations/en_US
rm messages.ini
mv messages.bak.ini messages.ini
```

**Comment by FKrafft** [comment]
> @do-web I guess the easiest is as follows. Can you access a command line for your Mautic installation?
> 
> ```
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Swiftmailer/Transport
> mv AmazonApiTransport.php AmazonApiTransport.bak.php
> wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php
> 
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Translations/en_US
> mv messages.ini messages.bak.ini
> wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Translations/en_US/messages.ini
> ```
> 
> Then test if it fixes the problem.
> If you want to revert back to the old code for whatever reason, simply run:
> 
> ```
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Swiftmailer/Transport
> rm AmazonApiTransport.php
> mv AmazonApiTransport.bak.php AmazonApiTransport.php
> 
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Translations/en_US
> rm messages.ini
> mv messages.bak.ini messages.ini
> ```
@dennisameling 
I will test today or tomorrow.  Your bug fix should work for 3.3.3 ?

**Comment by FKrafft** [comment]
> > @do-web I guess the easiest is as follows. Can you access a command line for your Mautic installation?
> > ```
> > (First go to the folder where you have installed Mautic!)
> > cd app/bundles/EmailBundle/Swiftmailer/Transport
> > mv AmazonApiTransport.php AmazonApiTransport.bak.php
> > wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php
> > 
> > (First go to the folder where you have installed Mautic!)
> > cd app/bundles/EmailBundle/Translations/en_US
> > mv messages.ini messages.bak.ini
> > wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Translations/en_US/messages.ini
> > ```
> > 
> > 
> > Then test if it fixes the problem.
> > If you want to revert back to the old code for whatever reason, simply run:
> > ```
> > (First go to the folder where you have installed Mautic!)
> > cd app/bundles/EmailBundle/Swiftmailer/Transport
> > rm AmazonApiTransport.php
> > mv AmazonApiTransport.bak.php AmazonApiTransport.php
> > 
> > (First go to the folder where you have installed Mautic!)
> > cd app/bundles/EmailBundle/Translations/en_US
> > rm messages.ini
> > mv messages.bak.ini messages.ini
> > ```
> 
> @dennisameling
> I will test today or tomorrow. Your bug fix should work for 3.3.3 ?



> @do-web I guess the easiest is as follows. Can you access a command line for your Mautic installation?
> 
> ```
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Swiftmailer/Transport
> mv AmazonApiTransport.php AmazonApiTransport.bak.php
> wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php
> 
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Translations/en_US
> mv messages.ini messages.bak.ini
> wget https://raw.githubusercontent.com/mautic/mautic/1dfdb01239e9244a8333314a3852fcd01ac163a5/app/bundles/EmailBundle/Translations/en_US/messages.ini
> ```
> 
> Then test if it fixes the problem.
> If you want to revert back to the old code for whatever reason, simply run:
> 
> ```
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Swiftmailer/Transport
> rm AmazonApiTransport.php
> mv AmazonApiTransport.bak.php AmazonApiTransport.php
> 
> (First go to the folder where you have installed Mautic!)
> cd app/bundles/EmailBundle/Translations/en_US
> rm messages.ini
> mv messages.bak.ini messages.ini
> ```

@dennisameling I tested with a small batch but still more than the AWS limit, it and it seems to work. 
Later this week I will do a test on a 50k+ email batch. I will let you know how it works.

**Comment by do-web** [comment]
i tested it with sending 7500 mails, and it worked, all mails were sent. In the SES console it also says that this number was sent.

But unfortunately the SES email bounce does not work, neither for SES API nor for SES SMTP. Both are set up on AWS

Bildschirmfoto 2021-05-11 um 18 08 05

**Comment by mabumusa1** [comment]
@do-web may you please send me the logs to see why it is not working? I tested multiple times and it worked as expected

**Comment by do-web** [comment]
> @do-web may you please send me the logs to see why it is not working? I tested multiple times and it worked as expected

@mabumusa1  Where i can see the logs?

**Comment by mabumusa1** [comment]
@do-web it is in the configuration menu under system information

**Comment by do-web** [comment]
> @do-web it is in the configuration menu under system information

```
[2021-05-11 11:22:15] mautic.ERROR: An exception occurred while executing 'UPDATE webhooks SET modified_by = ?, modified_by_user = ? WHERE id = ?' with params [null, " ", 1]:  SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction [] []
[2021-05-11 11:22:15] mautic.CRITICAL: Uncaught PHP Exception Doctrine\ORM\ORMException: "The EntityManager is closed." at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php line 144 {"exception":"[object] (Doctrine\\ORM\\ORMException(code: 0): The EntityManager is closed. at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php:144)"} []
[2021-05-11 11:22:15] mautic.ERROR: An exception occurred while executing 'UPDATE webhooks SET modified_by = ?, modified_by_user = ? WHERE id = ?' with params [null, " ", 1]:  SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction [] []
[2021-05-11 11:22:15] mautic.CRITICAL: Uncaught PHP Exception Doctrine\ORM\ORMException: "The EntityManager is closed." at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php line 144 {"exception":"[object] (Doctrine\\ORM\\ORMException(code: 0): The EntityManager is closed. at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php:144)"} []
[2021-05-11 11:50:03] mautic.ERROR: An exception occurred while executing 'UPDATE webhooks SET modified_by = ?, modified_by_user = ? WHERE id = ?' with params [null, " ", 1]:  SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction [] []
[2021-05-11 11:50:03] mautic.CRITICAL: Uncaught PHP Exception Doctrine\ORM\ORMException: "The EntityManager is closed." at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php line 144 {"exception":"[object] (Doctrine\\ORM\\ORMException(code: 0): The EntityManager is closed. at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php:144)"} []
[2021-05-11 17:54:27] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\HttpException: "AmazonCallback: Invalid JSON Payload" at /var/www/html/app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php line 77 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\HttpException(code: 0): AmazonCallback: Invalid JSON Payload at /var/www/html/app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:77)"} []
```

**Comment by do-web** [comment]
ok my mistake, I had deposited the SES topic with the wrong email.

**Comment by FKrafft** [comment]
Ok I tested sending and the numbers are inconsistent using the API. 
FYI: AWS limit 14/second.
I did not limit the batch size in mautic to 14 / second when initiating the send job. I used 100, assuming mautic would limit the send rate. 

Mautic: 35220 Sent
AWS: 32827 Sent

**Comment by FKrafft** [comment]
@dennisameling what do you need from me. Something that I should check/control?

**Comment by FKrafft** [comment]
@danielcherrington 
I found that it says "en_US". 
app/bundles/EmailBundle/Translations/en_US

The system, user, and email language were set to a non-English language. I will see if we can do another test with all set to English to see if this was the reason for the update not helping.

**Comment by FKrafft** [comment]
> @danielcherrington
> I found that it says "en_US".
> app/bundles/EmailBundle/Translations/en_US
> 
> The system, user, and email language were set to a non-English language. I will see if we can do another test with all set to English to see if this was the reason for the update not helping.

@danielcherrington 

This week's test worked out well, all 35k emails were sent out. 
1. Set language to English in the email template.
2. Set batch size the same as AWS SES limit. 

I might be wrong but it is my guess that the API batch transfer was slower than the limit of SES (14 emails/sec). And therefore the limit you hardcoded in the PHP file was never in use. If my memory is right, I saw some line about asking AWS for the max emails per second. 

I'll run another test this week, will set the batch size to 2x of the SES limit.

FYI, before it was set to 100 / batch, and SES limit was 14/second

**Comment by FKrafft** [comment]
Today we tested sending batches of size 2x of AWS allowance. That is 14*2 = 28 emails per bath. It did send all the emails. 
Total emails sent was 35,278

We will test 3x on Sunday.

**Comment by FKrafft** [comment]
When we increased the batch size to 3x the was ses allowance (14/second) then less was delivered. 
Out of the emailing list of 35k, only 28.8k was sent.

So the code is not limiting the transfer. 

@dennisameling  @mabumusa1 
Suggestions? 
The PHP files were updated according to your suggestion. 
Is there a log somewhere associated with AWS that I can download and post?

**Comment by mabumusa1** [comment]
@RCheesley I want to pick up this PR again, can you help me get it tested or point me to people who might be able to help on testing it?

**Comment by RCheesley** [comment]
@mabumusa1 I'd suggest making a request in the Product Team channel to see if folks are able to test.

**Comment by alexhammerschmied** [comment]
Are there any news on this? 
It seem still to cause some problems with the api

**Comment by mabumusa1** [comment]
> Are there any news on this? It seem still to cause some problems with the api

@alexhammerschmied please test and review this PR. I think it solve the speed issue

**Comment by mabumusa1** [comment]
To give a quick update on this PR, I tested the code after rebasing to `4.x`  I think the problem we are facing here is more general and not related to SES alone (although this PR solve issues with SES transport), below are the some of the results

**On Immediate Mode**:

When Mautic is configured to use Immediate mode there are multiple factors that impact sending speed, not just the negation between the server and the mail server to accept the API call, there is also the processing time within Mautic to create the batch (the larger the batch the more time it needs to process) 

In the screenshot below, we can see the request takes from 35-42 seconds from the ajax request is made until the email is sent (this time has multiple factors like connection speed, server capacity, and request processing time), in this configuration I matched SES request limit of 160 emails/second and the batch size to be 160 emails 

image

I checked the code line by line using Xdebug to see how the values are being transmitted and if concurrency is in place, everything worked as expected. SES did not raise an issue and emails were delivered as expected.


**On Queue Mode**:

The problem is bigger because we spool the fils one by one and we handle them one by one, so no matter what you do you are still handling files one by one, and speed limits are not respected.


**Suggested Solutions**
- Implement another spool that does not depend on files but depends on arrays where we save tokens, the parts of the email, and then send an API call. **this will make SMTP break**
- Make immediate mode faster in handling emails to match the send speed


**Update**
It took 16.6 Minutes to send 5000 emails with SES account limited with 160 emails/second
image

**Another Update**
It took 2 minutes to send 5000 on a server using a cron job

`php bin/console mautic:broadcasts:send --channel=email --id=1 --limit=500 --batch=500 -v`

Screenshot 2021-11-20 10:36:51

**Comment by cswainhart** [comment]
We used the PR in our production environment and it worked as expected.

**Comment by alexhammerschmied** [comment]
I just sent 25100 emails via broadcast cronjob. Planned published. So entirely in the background. 
But it took 1h45 minutes to process all emails. This is very long. « these emails are not yet in the Amazon SES Console Log for the specific configuration set (about 20 Minutes later).

At the same time i sent 11500 E-Mails for another client in the foreground (browser and click send button) with batch size 50: This took around 7 minutes. « these emails were about two minutes after sending the whole batch also in the Amazon SES Console Log for the specific configuration set

I'll send another batch tomorrow with the 25000 contacts. Again in the background with the batch flag and batchsize 50 mautic:broadcasts:send -b 50 let's see what happens.

**Comment by dsp76** [comment]
From what I see in the code, this checks the 24h sending limit with SES, but not the speed limitation given by Amazon SES  of x mails per sec?

**Comment by RCheesley** [comment]
@mabumusa1 I think this PR needs a bit more work based on the comment above, so I am going to remove the milestone to give us some more time. When it's ready we can consider it for a future release.

**Comment by alexhammerschmied** [comment]
I just sent another broadcast with 25100 mails. This time i sent it with the flag -b 50 (because 50mails/second is the current sendspeed for this project).
Now it worked like a charm. Planned the mail, broadcast cronjob ran, mails were sent within 6 minutes and a few seconds. 
Amazing. 
So for me, the code works but the batch size needed to be specified to send in fullspeed.

**Comment by dsp76** [comment]
@alexhammerschmied this would work if at the same time the batch can be fixed to 1 batch per second. It looks like it happens to be the case for you, but a quicker server maybe sends more batches per second?

**Comment by kuzmany** [comment]
@alexhammerschmied `mautic:broadcast:send -b 50` does not send in batch of 50 (description and docs are wrong)
If you want send email in batch of 50  you have to use 

`mautic:broadcast:send --limit=50 --batch=1`

I've open PR to modify description of command. https://github.com/mautic/mautic/pull/10622

**Comment by dsp76** [comment]
@kuzmany --batch=1 means, its sending then 50 messages per cron job run?

**Comment by hounw** [comment]
From what I gathered this issue is not yet fixed correct? Any known workarounds in the meantime? this is urgent for us...

**Comment by dsp76** [comment]
@hounw it seems not to address the limit per second, but the limit per 24h.

**Comment by mabumusa1** [comment]
I am closing this PR in favor of #11180 , everything has changed on that PR so this one has no meaning anymore.

All the issues mentioned here are fixed there

**Pull request closed without merging**